File: 0046.permutations/solution.go
package leetcode0046

func permute(nums []int) [][]int {
	return dfs(0, len(nums), nums)
}

func dfs(first, length int, nums []int) (ans [][]int) {
	// fixme
	if first == length {
		ans = append(ans, nums)
	}
	for i := first; i < length; i++ {
		nums[i], nums[first] = nums[i], nums[first]
		dfs(first+1, length, nums)
		nums[i], nums[first] = nums[i], nums[first]
	}
	return ans
}

File: 0046.permutations/solution.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def dfs(first):
            if first == length:
                ans.append(nums[:])
            for i in range(first, length):
                nums[i], nums[first] = nums[first], nums[i]
                dfs(first + 1)  # dfs
                nums[i], nums[first] = nums[first], nums[i]  # reduction

        ans = []
        length = len(nums)
        dfs(0)
        return ans
